arangod/Cluster/ClusterInfo.cpp

```diff
@@ -4442,6 +4446,98 @@ Result ClusterInfo::finishModifyingAnalyzerCoordinator(
   return Result(TRI_ERROR_NO_ERROR);
 }
 
+void ClusterInfo::initMetricsState() {
+  auto rebootId = ServerState::instance()->getRebootId().value();
+  auto serverId = ServerState::instance()->getId();
+
+  VPackBuilder builderRebootId;
+  builderRebootId.add(VPackValue{rebootId});
+  VPackBuilder builderServerId;
+  builderServerId.add(VPackValue{serverId});
+
+  AgencyWriteTransaction const write{
+      {{kMetricsRebootId, AgencyValueOperationType::SET,
+        builderRebootId.slice()},
+       {kMetricsServerId, AgencyValueOperationType::SET,
+        builderServerId.slice()}},
+      {{kMetricsRebootId, AgencyPrecondition::Type::EMPTY, true},
+       {kMetricsServerId, AgencyPrecondition::Type::EMPTY, true}}};
+  AgencyComm ac{_server};
+  while (!server().isStopping()) {
+    auto const r = ac.sendTransactionWithFailover(write);
+    if (r.successful() ||
+        r.httpCode() == rest::ResponseCode::PRECONDITION_FAILED) {
+      return;
+    }
+    LOG_TOPIC("bfdc3", WARN, Logger::CLUSTER)
+        << "Failed to self-propose leader with httpCode: " << r.httpCode();
+  }
+}
+
+ClusterInfo::MetricsState ClusterInfo::getMetricsState(bool wantLeader) {
+  auto& ac = _server.getFeature<ClusterFeature>().agencyCache();
+  auto [result, index] = ac.read({AgencyCommHelper::path(kMetricsServerId),
+                                  AgencyCommHelper::path(kMetricsRebootId)});
+  auto data = result->slice().at(0).get(std::initializer_list<std::string_view>{
+      AgencyCommHelper::path(), "Plan", "Metrics"});
+  auto leaderRebootId = data.get("RebootId").getNumber<uint64_t>();
+  auto leaderServerId = data.get("ServerId").stringView();
+  auto ourRebootId = ServerState::instance()->getRebootId().value();
+  auto ourServerId = ServerState::instance()->getId();
+  if (ourRebootId == leaderRebootId && ourServerId == leaderServerId) {
+    TRI_ASSERT(_metricsGuard.empty());
```

Comment:
True, but I think we don't need to explicit remove it here, because:
`_metricsGuard` dtor in that case doesn't do something useful (just check in reboot tracker callbacks that our leader propose callback have already called)

Reply:
Done and made rebase to devel

Reply:
The `_metricsGuard` destructor unregisters the callback with the RebootTracker, as far as I understand. So we should better call this to get rid of the registration.

Reply:
I don't want early call it because it possible only under `if (want leader)`, otherwise race cond.
I think it's ugly. So I prefer lazy useless guard destroy

Reply:
No, I think you are confusing the two callbacks which are in play here. There is one callback which you install to be called, if the server given changes its rebootId. You have no direct access to this callback, it is stored in the RebootTracker. Then the method `callMeOnChange` returns a second callback, called a `CallbackGuard`. This is something else. It is a closure which essentially does this:
```
  auto callbackGuard = CallbackGuard([this, peerState, callbackId]() {
    unregisterCallback(peerState, callbackId);
  });
```
The purpose of this is to unregister your callback, if the guard goes out of scope. In particular, calling this has the side effect of unregistering your callback.
If we become leader, then yes, our callback has been called, but I think it is still the right thing to do to now destruct the `CallbackGuard`, which will unregister a callback, which is only necessary on followers. So the newly elected leader should remove it explicitly, if only to make it clear to the reader of the code that this callback is no longer needed.

Reply:
> If we become leader, then yes, our callback has been called, but I think it is still the right thing to do to now destruct the CallbackGuard, which will unregister a callback, which is only necessary on followers. So the newly elected leader should remove it explicitly, if only to make it clear to the reader of the code that this callback is no longer needed.

As I wrote it just choice for readability.
Because we can remove it only `if wantLeader`, so I think not necessary do it here, additional code, that doesn't make something useful